Documentation/teaching/labs/device_drivers.rst

```diff
@@ -948,46 +948,82 @@ For this:
    1. Implement the ioctl function in the driver.
    2. We need to use ``user/so2_cdev_test.c`` to call the
       ioctl function with the appropriate parameters.
+   3. To test, we will use an user-space program (``user/so2_cdev_test.c``)
+      which will call the ``ioctl`` function with the required arguments.
+
+.. note:: The macro ``MY_IOCTL_PRINT`` is defined in the file ``include/so2_cdev.h``,
+          which is shared between the kernel module and the user-space program.
 
-.. note:: The macro definition ``MY_IOCTL_PRINT`` is defined in the ``include/so2_cdev.h`` file
           Read the `ioctl`_ section in the lab.
 
-.. note:: Because we need to compile the program for qemu machine which is 32 bit, if your host is 64 bit
-          then you need to install ``gcc-multilib`` package.
+.. note:: The userspace code is compiled automatically at ``make build`` and
+          copied at ``make copy``.
+
+          Because we need to compile the program for qemu machine which is 32 bit,
+          if your host is 64 bit then you need to install ``gcc-multilib`` package.
+
+Extra Exercises
+===============
+
+Ioctl with messaging
+--------------------
+
+Add two ioctl operations to modify the message associated with the
+driver. Use fixed-length buffer ( BUFFER_SIZE ).
+
+  1. Add the ``ioctl`` function from the driver the following operations:
+     * ``MY_IOCTL_SET_BUFFER`` for writing a message to the device;
+     * ``MY_IOCTL_GET_BUFFER`` to read a message from your device.
+  2. Change the user-space program to allow testing.
+
+.. note:: Read the ioctl sections and Access to the address space of the lab process.
+
+Ioctl with waiting queues
+-------------------------
+
+Add two ioctl operations to the device driver for queuing.
+
+   1. Add the ``ioctl`` function from the driver the following operations:
+      * ``MY_IOCTL_DOWN`` to add the process to a queue;
+      * ``MY_IOCTL_UP`` to remove the process from a queue.
+   2. Fill the device structure with a ``wait_queue_head_t`` field and a
+      ``wait_queue_head_t`` flag.
+   3. Do not forget to initialize the wait queue and flag.
+   4. Remove exclusive access condition from previous exercise
+   5. Change the user-space program to allow testing.
 
-.. Extra
-   -----
+When the process is added to the queue, it will remain blocked in execution; To
+run the queue command open a new console in the virtual machine with Alt+F2 ;
+You can return to the previous console with Alt+F1 . If you're connected via
+SSH to the virtual machine, open a new console.
 
-   Ioctl with messaging
-   --------------------
+.. note:: Read the ioctl and Synchronization sections - waiting queues in the lab.
```

Comment:
Don't forget to add hyperlinks to the sections.

Reply:
Fixed.